File: app/src/main/java/xyz/volgoak/wordlearning/model/WordsViewModel.java
package xyz.volgoak.wordlearning.model;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;
import android.util.Log;

import com.annimon.stream.Stream;

import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import javax.inject.Inject;

import io.fabric.sdk.android.services.concurrency.AsyncTask;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import timber.log.Timber;
import xyz.volgoak.wordlearning.WordsApp;
import xyz.volgoak.wordlearning.data.DataProvider;
import xyz.volgoak.wordlearning.data.DatabaseContract;
import xyz.volgoak.wordlearning.entities.Set;
import xyz.volgoak.wordlearning.entities.Theme;
import xyz.volgoak.wordlearning.entities.Word;

/**
 * Created by alex on 2/13/18.
 */

public class WordsViewModel extends ViewModel {

    @Inject
    DataProvider provider;

    private List<Set> sets;
    private List<Theme> themes;
    private MutableLiveData<List<Set>> setsLiveData = new MutableLiveData<>();
    private MutableLiveData<List<Theme>> themesLiveData = new MutableLiveData<>();

    private MutableLiveData<Set> selectedSetData = new MutableLiveData<>();
    private MutableLiveData<List<Word>> selectedSetWordsData = new MutableLiveData<>();

    private Disposable wordsDisposable;
    private Disposable setDisposable;

    private ExecutorService executor = Executors.newSingleThreadExecutor();


    public WordsViewModel() {
        WordsApp.getsComponent().inject(this);
    }

    private boolean isSetLoaded;

    public LiveData<List<Set>> getSets(String theme) {
        if (sets == null || sets.size() == 0) {
            executor.submit(() -> {
                sets = provider.getAllSets();
                List<Set> themed = Stream.of(sets)
                        .filter(set -> set.getThemeCodes().contains(theme))
                        .toList();
                setsLiveData.postValue(themed);
            });
        }
        return setsLiveData;
    }

    public LiveData<List<Theme>> getThemes() {
        if (themes == null || themes.size() == 0) {
            executor.submit(() -> {
                themes = provider.getAllThemes();
                themesLiveData.postValue(themes);
            });
        }

        return themesLiveData;
    }

    public LiveData<List<Word>> getWordsForSet() {
        return selectedSetWordsData;
    }

    public LiveData<Set> getCurrentSet() {
        return selectedSetData;
    }

    public LiveData<Boolean> changeSet(long setId) {
        isSetLoaded = true;
        MutableLiveData<Boolean> loadedCallback = new MutableLiveData<>();

        if(wordsDisposable != null && !wordsDisposable.isDisposed()) {
            wordsDisposable.dispose();
        }

        if(setDisposable != null && !setDisposable.isDisposed()) {
            setDisposable.dispose();
        }

        wordsDisposable = provider.getWordsBySetIdRx(setId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(list -> {
                    Timber.d("words from flowable");
                    selectedSetWordsData.setValue(list);
                }, Timber::e);

        setDisposable = provider.getSetById(setId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(set -> {
                    loadedCallback.setValue(true);
                    selectedSetData.setValue(set);
                    Timber.d("set from flowable %1$s", set.getName());
                }, Timber::e);

        return loadedCallback;
    }

    public LiveData<Boolean> changeToDictionary() {
        if(wordsDisposable != null && !wordsDisposable.isDisposed()) {
            wordsDisposable.dispose();
        }

        if(setDisposable != null && !setDisposable.isDisposed()) {
            setDisposable.dispose();
        }

        MutableLiveData<Boolean> loadedCallback = new MutableLiveData<>();

        wordsDisposable = provider.getDictionaryWordsFlowable()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(list -> {
                    loadedCallback.postValue(true);
                    selectedSetWordsData.postValue(list);
                }, Timber::e);

        return loadedCallback;
    }

    public LiveData<Integer> changeCurrentSetStatus() {
        MutableLiveData<Integer> statusLive = new MutableLiveData<>();
        Set current = selectedSetData.getValue();
        if (current != null) {
            int newStatus = current.getStatus() == DatabaseContract.Sets.IN_DICTIONARY ?
                    DatabaseContract.Sets.OUT_OF_DICTIONARY : DatabaseContract.Sets.IN_DICTIONARY;

            current.setStatus(newStatus);
            executor.submit(() -> {
                provider.updateSetStatus(current);
                statusLive.postValue(newStatus);
            });
        }
        return statusLive;
    }

    public void resetCurrentSetProgress() {
        Set current = selectedSetData.getValue();
        if (current != null) {
            executor.submit(() -> provider.resetSetProgress(current));
        }
    }

    public void updateWords(Word[] words) {
        executor.submit(() -> provider.updateWords(words));
    }

    public void resetWordsProgress(List<Integer> positions) {
        Set set = selectedSetData.getValue();
        if (set != null && positions.size() != 0) {
            executor.submit(() -> {
                Word[] wordsArray = new Word[positions.size()];
                List<Word> words = provider.getWordsBySetId(set.getId());
                for (int a = 0; a < positions.size(); a++) {
                    Word word = words.get(positions.get(a));
                    word.resetProgress();
                    wordsArray[a] = word;
                }
                provider.updateWords(wordsArray);
            });
        }
    }

    public void changeWordsStatus(List<Integer> positions, int status) {
        Set set = selectedSetData.getValue();
        if (set != null && positions.size() != 0) {
            executor.submit(() -> {
                Word[] wordsArray = new Word[positions.size()];
                List<Word> words = provider.getWordsBySetId(set.getId());
                for (int a = 0; a < positions.size(); a++) {
                    Word word = words.get(positions.get(a));
                    word.setStatus(status);
                    wordsArray[a] = word;
                }
                provider.updateWords(wordsArray);
            });
        }
    }

    public void changeTheme(String theme) {
        executor.submit(() -> {
            List<Set> themeSet = Stream.of(sets).filter(set -> set.getThemeCodes().contains(theme)).toList();
            setsLiveData.postValue(themeSet);
        });
    }

    public void updateSetStatus(Set set) {
        executor.submit(() -> provider.updateSetStatus(set));
    }

    public void insertWord(Word newWord) {
        executor.submit(() -> provider.insertWord(newWord));
    }

    public void deleteOrHideWord(Word word) {
        word.setStatus(DatabaseContract.Words.OUT_OF_DICTIONARY);
        AsyncTask.execute(() -> provider.updateWords(word));
    }

    public boolean isSetLoaded() {
        return isSetLoaded;
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        executor.shutdown();
    }
}

File: app/src/bottom_nav/java/xyz/volgoak/wordlearning/screens/set/viewModel/WordsViewModel.kt
package xyz.volgoak.wordlearning.screens.set.viewModel

import android.arch.lifecycle.*

import com.attiladroid.data.DataProvider
import com.attiladroid.data.DataContract
import com.attiladroid.data.entities.Word
import com.attiladroid.data.entities.Set
import java.util.concurrent.Executors

import javax.inject.Inject

import io.fabric.sdk.android.services.concurrency.AsyncTask
import xyz.volgoak.wordlearning.R.string.word
import xyz.volgoak.wordlearning.WordsApp
import xyz.volgoak.wordlearning.utils.SingleLiveEvent


/**
 * Created by alex on 2/13/18.
 */

class WordsViewModel(val id: Long) : ViewModel() {

    @Inject
    lateinit var provider: DataProvider

    val setData by lazy { provider.getSetById(id)}
    val wordsData by lazy { provider.getWordsBySetIdLD(id)}
    val startCardsData = SingleLiveEvent<Int>()

    private val executor = Executors.newSingleThreadExecutor()

    init {
        WordsApp.getsComponent().inject(this)
    }

    fun changeCurrentSetStatus() {
        val current = setData.value
        if (current != null) {
            val newStatus = if (current.status == DataContract.Sets.IN_DICTIONARY)
                DataContract.Sets.OUT_OF_DICTIONARY
            else
                DataContract.Sets.IN_DICTIONARY

            current.status = newStatus
            executor.submit {
                provider.updateSetStatus(current)
            }
        }
    }

    fun resetCurrentSetProgress() {
        val current = setData.value
        if (current != null) {
            executor.submit { provider.resetSetProgress(current) }
        }
    }

    fun updateWords(words: Array<Word>) {
        executor.submit { provider.updateWords(*words) }
    }

    fun resetWordsProgress(positions: List<Int>) {
        val set = setData.value
        if (set != null && positions.isNotEmpty()) {
            executor.submit {
                val wordsList = mutableListOf<Word>()
                val words = provider.getWordsBySetId(set.id)
                for (a in positions.indices) {
                    val word = words[positions[a]]
                    word.resetProgress()
                    wordsList.add(word)
                }
                provider.updateWords(wordsList)
            }
        }
    }

    fun changeWordsStatus(positions: List<Int>, status: Int) {
        val set = setData.value
        if (set != null && positions.isNotEmpty()) {
            executor.submit {
                val wordsList = mutableListOf<Word>()
                val words = provider.getWordsBySetId(set.id)
                for (a in positions.indices) {
                    val word = words[positions[a]]
                    word.status = status
                    wordsList.add(word)
                }
                provider.updateWords(wordsList)
            }
        }
    }

    fun updateSetStatus(set: Set) {
        executor.submit { provider.updateSetStatus(set) }
    }

    fun insertWord(newWord: Word) {
        executor.submit<Long> { provider!!.insertWord(newWord) }
    }

    fun deleteOrHideWord(word: Word) {
        word.status = DataContract.Words.OUT_OF_DICTIONARY
        AsyncTask.execute { provider!!.updateWords(word) }
    }

    override fun onCleared() {
        super.onCleared()
        executor.shutdown()
    }

    @Suppress("UNCHECKED_CAST")
    class Factory(val id: Long): ViewModelProvider.NewInstanceFactory() {
        override fun <T : ViewModel?> create(modelClass: Class<T>): T {
            return WordsViewModel(id) as T
        }
    }
}
